src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -23,6 +26,98 @@ public static class Renamer
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentName == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentName));
+            }
+
+            using var _ = ArrayBuilder<RenameDocumentAction>.GetInstance(out var actions);
+
+            if (!newDocumentName.Equals(document.Name))
+            {
+                var renameAction = await RenameSymbolDocumentAction.TryCreateAsync(document, newDocumentName, cancellationToken).ConfigureAwait(false);
+
+                if (renameAction is object)
+                {
+                    actions.Add(renameAction);
+                }
+            }
+
+            return new RenameDocumentActionSet(
+                actions.ToImmutable(),
+                document.Project.Id,
+                document.Id,
+                newDocumentName,
+                document.Folders,
+                optionSet);
+        }
+
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithFolders(IEnumerable{string})" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="SyncNamespaceDocumentAction"/> that will sync the namespace(s) of the document to match the document folders
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentFoldersAsync(
+            Document document,
+            IReadOnlyList<string> newFolders,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newFolders == null)
+            {
+                throw new ArgumentNullException(nameof(newFolders));
+            }
+
+            using var _ = ArrayBuilder<RenameDocumentAction>.GetInstance(out var actions);
+
+            if (!newFolders.SequenceEqual(document.Folders))
+            {
+                var action = await SyncNamespaceDocumentAction.TryCreateAsync(document, newFolders, cancellationToken).ConfigureAwait(false);
+
+                if (action is object)
+                {
+                    actions.Add(action);
+                }
+            }
+
+            return new RenameDocumentActionSet(
+                actions.ToImmutable(),
+                document.Project.Id,
+                document.Id,
+                document.Name,
+                newFolders,
+                optionSet);
+        }
```

Comment:
why have the two entrypoints, instead of one entrypoint that can take in a newName, and optional newFolders?  Clients can pass whatever they want for that, and we can see about both renaming and sync-ing folders.  You could then potentially get 0-2 actions back.  Taht would make a lot more sense to me and would have one single APi entrypoint for people to use, and would motivate why we need to support multiple actions.

Reply:
What should the suggested single API design look like? Should both folders and document name be required? Optional, but throw error if both aren't supplied?

Reply:
i would make the new doc-name non-optional, and hte folders optional.  but doc that the new doc-name can be null (or hte same value as hte original name to not change it).